observability/tracing.go

```diff
@@ -0,0 +1,63 @@
+// The observability package provides utilities for the Kiali server
+// to instrument itself with observability tools such as tracing to provide
+// better insights into server performance.
+package observability
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/jaeger"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+
+	"github.com/kiali/kiali/config"
+)
+
+const (
+	// TracingService is the name of the kiali tracer service.
+	TracingService = "kiali"
```

Comment:
I don't know if it is OK for this to be fixed to `kiali` or not, but I do want to draw your attention to the `deployment.instance_name` feature. You can have a Kiali installed with different names, and you can have multiple Kialis deployed in the same namespace. So I don't know if this `TracingService` name should be this `deployment.instance_name` (the default value is `kiali`) or if it is OK for this to be a constant `kiali`.

See docs on this:
* https://github.com/kiali/kiali-operator/blob/v1.44.0/deploy/kiali/kiali_cr.yaml#L358-L373
* https://kiali.io/docs/installation/deployment-options/#kiali-instance-name

Reply:
If `deployment.instance_name` determines the "workload" name for the app then this is probably what we want to use as the [hostname](https://github.com/kiali/kiali/pull/4532/commits/8837d47a8694fc801d028c979085a240ff9cd579#diff-296a65271be18baa8a1fe01a82d31a85bcb3b223249922564a42d7db3f999ba8R48). `TracingService` is more along the lines of `canonical_service` or the kube service name. I can't really link to the specific open telemetry constants but here's an excerpt:
> 	// Logical name of the service.
	//
	// Type: string
	// Required: Always
	// Stability: stable
	// Examples: 'shoppingcart'

from https://pkg.go.dev/go.opentelemetry.io/otel/semconv/v1.7.0#pkg-constants.

I'll update the hostname resource with the `deployment.instance_name` field as we want that to match the name of the workload since kiali uses that to determine if the trace matches the workload: https://github.com/kiali/kiali/blob/master/business/jaeger.go#L275-L286.

Reply:
Ah I was hoping `deployment.instance_name` would be available on the kiali config but it doesn't appear to be. Ideally the workload name would get pulled from the pod annotations and passed down through the downward api similar to how the pod name is set. That would be less error prone than assuming the workload will always be `kiali`. That being said, it looks like the app name when deploying with the operator will always be `kiali`: https://github.com/kiali/kiali-operator/blob/v1.44.0/roles/default/kiali-deploy/templates/kubernetes/deployment.yaml#L11 so I think it's fine to assume here the workload will just be `kiali`.

Reply:
I'm surprised it isn't available... everything in the Kiali CR will get copied to the Kiali ConfigMap, even the deployment.* settings. See: https://github.com/kiali/kiali-operator/blob/master/roles/default/kiali-deploy/templates/openshift/configmap.yaml#L9

Same should be true with the server helm chart, though interestingly, the instance_name is set to the helm chart full name (which I thought should be the instance name, but maybe not?): https://github.com/kiali/helm-charts/blob/master/kiali-server/templates/configmap.yaml#L19

Anyway, if its fine for the traces to use the app name label, then it will always be kiali. I just thought it would be weird that you could have multiple instances installed and no way of knowing WHICH trace came from WHICH instance. But maybe that doesn't matter?

Reply:
> I'm surprised it isn't available... everything in the Kiali CR will get copied to the Kiali ConfigMap, even the deployment.* settings. See: https://github.com/kiali/kiali-operator/blob/master/roles/default/kiali-deploy/templates/openshift/configmap.yaml#L9

It looks like it's there in the configmap yaml it's just doesn't exist on the config go struct: https://github.com/kiali/kiali/blob/master/config/config.go#L325-L330. That's an easy fix.

Reply:
![Screenshot from 2022-01-04 12-06-49](https://user-images.githubusercontent.com/6226732/148096561-053a3494-dce0-4333-bfb0-5524e288fd35.png)

There we go